File: apps/cli/src/test/lib/format.test.ts
import { formatBytes } from "../../lib/format";

describe("formatBytes", () => {
    describe("with default options (binary units)", () => {
        it("should format 0 bytes", () => {
            expect(formatBytes(0)).toBe("0 Bytes");
        });

        it("should format bytes under 1 KiB", () => {
            expect(formatBytes(100)).toBe("100 Bytes");
            expect(formatBytes(1023)).toBe("1,023 Bytes");
        });

        it("should format exactly 1 KiB", () => {
            expect(formatBytes(1024)).toBe("1 KiB");
        });

        it("should format KiB values", () => {
            expect(formatBytes(1536)).toBe("1.5 KiB");
            expect(formatBytes(2048)).toBe("2 KiB");
            expect(formatBytes(10240)).toBe("10 KiB");
            expect(formatBytes(102400)).toBe("100 KiB");
        });

        it("should format MiB values", () => {
            expect(formatBytes(1048576)).toBe("1 MiB");
            expect(formatBytes(1572864)).toBe("1.5 MiB");
            expect(formatBytes(10485760)).toBe("10 MiB");
            expect(formatBytes(104857600)).toBe("100 MiB");
        });

        it("should format GiB values", () => {
            expect(formatBytes(1073741824)).toBe("1 GiB");
            expect(formatBytes(1610612736)).toBe("1.5 GiB");
            expect(formatBytes(10737418240)).toBe("10 GiB");
            expect(formatBytes(107374182400)).toBe("100 GiB");
        });

        it("should format TiB values", () => {
            expect(formatBytes(1099511627776)).toBe("1 TiB");
            expect(formatBytes(1649267441664)).toBe("1.5 TiB");
        });

        it("should handle intelligent decimal places", () => {
            // Values >= 100 should have no decimals
            expect(formatBytes(104857600)).toBe("100 MiB");
            expect(formatBytes(107374182400)).toBe("100 GiB");
            
            // Values 10-99 should have up to 1 decimal
            expect(formatBytes(12582912)).toBe("12 MiB");
            expect(formatBytes(13107200)).toBe("12.5 MiB");
            
            // Values < 10 should have up to 2 decimals
            expect(formatBytes(5242880)).toBe("5 MiB");
            expect(formatBytes(5767168)).toBe("5.5 MiB");
            expect(formatBytes(2359296)).toBe("2.25 MiB");
        });
    });

    describe("with decimal units", () => {
        const decimalOptions = { binary: false };

        it("should format exactly 1 KB", () => {
            expect(formatBytes(1000, decimalOptions)).toBe("1 KB");
        });

        it("should format KB values", () => {
            expect(formatBytes(1500, decimalOptions)).toBe("1.5 KB");
            expect(formatBytes(10000, decimalOptions)).toBe("10 KB");
            expect(formatBytes(100000, decimalOptions)).toBe("100 KB");
        });

        it("should format MB values", () => {
            expect(formatBytes(1000000, decimalOptions)).toBe("1 MB");
            expect(formatBytes(1500000, decimalOptions)).toBe("1.5 MB");
            expect(formatBytes(10000000, decimalOptions)).toBe("10 MB");
        });

        it("should format GB values", () => {
            expect(formatBytes(1000000000, decimalOptions)).toBe("1 GB");
            expect(formatBytes(1500000000, decimalOptions)).toBe("1.5 GB");
        });
    });

    describe("with custom locale", () => {
        it("should format with German locale", () => {
            expect(formatBytes(1536, { locale: "de-DE" })).toBe("1,5 KiB");
            expect(formatBytes(1234567890, { locale: "de-DE" })).toBe("1,15 GiB");
        });

        it("should format with French locale", () => {
            expect(formatBytes(1536, { locale: "fr-FR" })).toBe("1,5 KiB");
        });
    });

    describe("with custom decimal places", () => {
        it("should respect custom decimal places for small values", () => {
            expect(formatBytes(2359296, { decimals: 3 })).toBe("2.25 MiB");
            expect(formatBytes(2411724, { decimals: 3 })).toBe("2.3 MiB");
            expect(formatBytes(2306867, { decimals: 3 })).toBe("2.2 MiB");
        });
    });
});

File: apps/cli-zig/src/test/format.test.zig
const std = @import("std");
const cli = @import("cli-zig");
const helpers = @import("test-helpers.zig");
const formatBytes = cli.format.formatBytes;

//
// Formats with the default options (binary units, 2 decimals).
//
fn format(allocator: std.mem.Allocator, bytes: f64) ![]const u8 {
    return formatBytes(allocator, bytes, cli.format.defaultFormatBytesOptions);
}

//
// Formats with decimal units.
//
fn formatDecimal(allocator: std.mem.Allocator, bytes: f64) ![]const u8 {
    return formatBytes(allocator, bytes, .{ .binary = false, .decimals = 2 });
}

test "should format 0 bytes" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    try std.testing.expectEqualStrings("0 Bytes", try format(arena.allocator(), 0));
}

test "should format bytes under 1 KiB" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    try std.testing.expectEqualStrings("100 Bytes", try format(arena.allocator(), 100));
    try std.testing.expectEqualStrings("1,023 Bytes", try format(arena.allocator(), 1023));
}

test "should format exactly 1 KiB" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    try std.testing.expectEqualStrings("1 KiB", try format(arena.allocator(), 1024));
}

test "should format KiB values" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    try std.testing.expectEqualStrings("1.5 KiB", try format(allocator, 1536));
    try std.testing.expectEqualStrings("2 KiB", try format(allocator, 2048));
    try std.testing.expectEqualStrings("10 KiB", try format(allocator, 10240));
    try std.testing.expectEqualStrings("100 KiB", try format(allocator, 102400));
}

test "should format MiB values" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    try std.testing.expectEqualStrings("1 MiB", try format(allocator, 1048576));
    try std.testing.expectEqualStrings("1.5 MiB", try format(allocator, 1572864));
    try std.testing.expectEqualStrings("10 MiB", try format(allocator, 10485760));
    try std.testing.expectEqualStrings("100 MiB", try format(allocator, 104857600));
}

test "should format GiB values" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    try std.testing.expectEqualStrings("1 GiB", try format(allocator, 1073741824));
    try std.testing.expectEqualStrings("1.5 GiB", try format(allocator, 1610612736));
    try std.testing.expectEqualStrings("10 GiB", try format(allocator, 10737418240));
    try std.testing.expectEqualStrings("100 GiB", try format(allocator, 107374182400));
}

test "should format TiB values" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    try std.testing.expectEqualStrings("1 TiB", try format(allocator, 1099511627776));
    try std.testing.expectEqualStrings("1.5 TiB", try format(allocator, 1649267441664));
}

test "should handle intelligent decimal places" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    try std.testing.expectEqualStrings("100 MiB", try format(allocator, 104857600));
    try std.testing.expectEqualStrings("100 GiB", try format(allocator, 107374182400));
    try std.testing.expectEqualStrings("12 MiB", try format(allocator, 12582912));
    try std.testing.expectEqualStrings("12.5 MiB", try format(allocator, 13107200));
    try std.testing.expectEqualStrings("5 MiB", try format(allocator, 5242880));
    try std.testing.expectEqualStrings("5.5 MiB", try format(allocator, 5767168));
    try std.testing.expectEqualStrings("2.25 MiB", try format(allocator, 2359296));
}

test "should format exactly 1 KB" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    try std.testing.expectEqualStrings("1 KB", try formatDecimal(arena.allocator(), 1000));
}

test "should format KB values" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    try std.testing.expectEqualStrings("1.5 KB", try formatDecimal(allocator, 1500));
    try std.testing.expectEqualStrings("10 KB", try formatDecimal(allocator, 10000));
    try std.testing.expectEqualStrings("100 KB", try formatDecimal(allocator, 100000));
}

test "should format MB values" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    try std.testing.expectEqualStrings("1 MB", try formatDecimal(allocator, 1000000));
    try std.testing.expectEqualStrings("1.5 MB", try formatDecimal(allocator, 1500000));
    try std.testing.expectEqualStrings("10 MB", try formatDecimal(allocator, 10000000));
}

test "should format GB values" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    try std.testing.expectEqualStrings("1 GB", try formatDecimal(allocator, 1000000000));
    try std.testing.expectEqualStrings("1.5 GB", try formatDecimal(allocator, 1500000000));
}

// Not ported: "should format with German locale", "should format with French locale" (only en-US is ported).

test "should respect custom decimal places for small values" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    try std.testing.expectEqualStrings("2.25 MiB", try formatBytes(allocator, 2359296, .{ .binary = true, .decimals = 3 }));
    try std.testing.expectEqualStrings("2.3 MiB", try formatBytes(allocator, 2411724, .{ .binary = true, .decimals = 3 }));
    try std.testing.expectEqualStrings("2.2 MiB", try formatBytes(allocator, 2306867, .{ .binary = true, .decimals = 3 }));
}

test "formatBytes matches the TypeScript output for every fixture value" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const fixture = try helpers.loadFixture(allocator, "format.json");
    for (fixture.array.items) |formatCase| {
        const bytes: f64 = @floatFromInt(helpers.intField(formatCase, "bytes"));
        const options: cli.format.IFormatBytesOptions = .{
            .binary = helpers.boolField(formatCase, "binary"),
            .decimals = @intCast(helpers.intField(formatCase, "decimals")),
        };
        const output = try formatBytes(allocator, bytes, options);
        std.testing.expectEqualStrings(helpers.stringField(formatCase, "output"), output) catch |err| {
            std.debug.print("bytes={d}\n", .{bytes});
            return err;
        };
    }
}

test "toLocaleString rounds half away from zero and groups thousands" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    try std.testing.expectEqualStrings("10.3", try cli.format.toLocaleString(allocator, 10.25, 1));
    try std.testing.expectEqualStrings("1.01", try cli.format.toLocaleString(allocator, 1.005, 2));
    try std.testing.expectEqualStrings("1.13", try cli.format.toLocaleString(allocator, 1.125, 2));
    try std.testing.expectEqualStrings("1,234,567", try cli.format.toLocaleString(allocator, 1234567, 0));
    try std.testing.expectEqualStrings("10", try cli.format.toLocaleString(allocator, 9.999, 2));
    try std.testing.expectEqualStrings("0", try cli.format.toLocaleString(allocator, 0, 2));
}
